phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/PhoenixMasterObserver.java

```diff
@@ -0,0 +1,187 @@
+     * We infer start and end key for the parent from the two daughters' start/end keys.
+     */
+    private String getParentPartitionId(Connection conn, String tableName, String streamName,
+                                        RegionInfo regionInfoA, RegionInfo regionInfoB)
+            throws SQLException {
+        byte[] inferredParentStartKey =
+                Bytes.compareTo(regionInfoA.getStartKey(), regionInfoB.getStartKey()) < 0 ?
+                regionInfoA.getStartKey() : regionInfoB.getStartKey();
+
+        byte[] inferredParentEndKey = (regionInfoB.getEndKey().length == 0 || regionInfoA.getEndKey().length == 0)
+                ? new byte[0]
+                : Bytes.compareTo(regionInfoA.getEndKey(), regionInfoB.getEndKey()) < 0
+                    ? regionInfoB.getEndKey()
+                    : regionInfoA.getEndKey();
+
+        StringBuilder qb = new StringBuilder(PARENT_PARTITION_QUERY);
+        if (inferredParentStartKey.length == 0) {
+            qb.append(" AND PARTITION_START_KEY IS NULL ");
+        } else {
+            qb.append(" AND PARTITION_START_KEY = ? ");
+        }
+        if (inferredParentEndKey.length == 0) {
+            qb.append(" AND PARTITION_END_KEY IS NULL ");
+        } else {
+            qb.append(" AND PARTITION_END_KEY = ? ");
+        }
+
+        PreparedStatement pstmt = conn.prepareStatement(qb.toString());
+        int index = 1;
+        pstmt.setString(index++, tableName);
+        pstmt.setString(index++, streamName);
+        if (inferredParentStartKey.length > 0) pstmt.setBytes(index++, inferredParentStartKey);
+        if (inferredParentEndKey.length > 0) pstmt.setBytes(index++, inferredParentEndKey);
+        LOGGER.info(pstmt.toString());
+        ResultSet rs = pstmt.executeQuery();
+        if (rs.next()) {
+            return rs.getString(1);
+        } else {
+            throw new SQLException("Could not find parent of the provided daughters.");
+        }
+    }
+
+    /**
+     * Insert partition metadata for a daughter region from the split.
+     */
+    private void upsertDaughterPartition(Connection conn, String tableName,
+                                         String streamName, String parentPartitionID,
+                                         RegionInfo regionInfo)
+            throws SQLException {
+        String partitionId = regionInfo.getEncodedName();
+        long startTime = regionInfo.getRegionId();
+        byte[] startKey = regionInfo.getStartKey();
+        byte[] endKey = regionInfo.getEndKey();
+        PreparedStatement pstmt = conn.prepareStatement(PARTITION_UPSERT_SQL);
+        pstmt.setString(1, tableName);
+        pstmt.setString(2, streamName);
+        pstmt.setString(3, partitionId);
+        pstmt.setString(4, parentPartitionID);
+        pstmt.setLong(5, startTime);
+        // endTime in not set when inserting a new partition
+        pstmt.setLong(6, -1L);
```

Comment:
Shall we keep it as `pstmt.setNull(6, Types.BIGINT)`?

Reply:
done